### ISSUE TITLE
Split `getRepresentation` into `getMembers` and `getRepresentation`?

### ISSUE THREAD

**Issue by michielbdejong** [RubenVerborgh/solid-server-ts#6]
Hi @RubenVerborgh,

As you know, I'm using the `ResourceStore` interface to implement components of a Solid server.
So to start off with the easy option, I started writing an in-memory resource store. I found that the first thing you have to do is put an `if` in `getRepresentation`, to distinguish between getting a container and getting a document. A document representation should be a Stream indeed, but if a container representation has to be a Stream at the ResourceStore level, then it needs to know about Turtle, which I think was probably not your intention.

So I added a `ResourceStore#getMembers` method (which does not know about Turtle and only returns a list of container member URLs) to resolve this. It occurred to me that this does mean that the store now knows about container tree structures, so maybe I'll rename it to `ResourceTreeStore`. In my in-memory implementations , only documents are stored, and containers are implied by the path prefixes of the documents, so it already knows about tree structures. But it would also be possible to have ResourceStore act as a dumb key-value store, where the next layer up explicitly updates the container member listings by writing Turtle to the store. I think I prefer the `ResourceTreeStore` approach, but wanted to check with you first if that's something you explicitly decided against when you designed the `ResourceStore` interface?

**Comment by RubenVerborgh** [comment]
TL;DR: I would create a `Container` interface with that method (and possibly some more).

----

I intended `ResourceStore` as the interface that handles resources and their representations in general, independent of what the resource represents (a type of container or something else). This has the following consequences:
- components using `ResourceStore` do not need to know how to construct representations or how to perform content negotiation
- we can easily apply the _Decorator_ pattern on `ResoureStore` to achieve various functions

As an example of the latter, I thought of using a `ResourceStore` for content negotiation. Suppose you have a `ResourceStore` that just reads files from disk, then you can put a `ResourceStore` in between that (for example) can translate between RDF formats. Or one that can generate virtual graphs. Etc.

Well, I guess all of that is still possible when you add an extra method. It's just that, when "listing things and representing them" becomes an additional responsibility, implementation get more complex. With the current abstraction, converting a folder listing to JSON-LD is the same a converting any other resource. With more complex interface, that becomes different.

I don't think `getRepresentation` should be _split_. I do think that there is a use case for `getMembers` (maybe even with a filter a argument, etc.). And `getRepresentation` could then internally (or not) invoke that method.
However, I would argue that this method (/ these methods) belong to a different interface called `Container` (subclass of `Resource`) or `ContainerStore`.

I think it's good to keep interfaces atomic, and `getMembers` might be a bit of a level breaker in that it returns a list of resource identifiers.

A counter-argument is that we currently have `addResource`. But maybe that method is in the wrong place, and we need a `getContainer` method somewhere that return a `Container` with `addResource` and `getResources`.

**Comment by michielbdejong** [comment]
> a getContainer method somewhere that return a Container with addResource and getResources.

Nice! I like it. Thanks!
